src/jolt/vm/pc.rs

```diff
@@ -0,0 +1,549 @@
+}
+
+impl<F: PrimeField> PCPolys<F> {
+  pub fn new(
+    a_ops: DensePolynomial<F>,
+    v_ops: DensePolynomial<F>,
+    v_mem: DensePolynomial<F>,
+    t_read: DensePolynomial<F>,
+    t_final: DensePolynomial<F>,
+    address_offset: usize,
+  ) -> Self {
+    // TODO(JOLT-48): DensePolynomial::merge should be parameterized with  Vec<&DensePolynomial> given it is read only. Avoids clones.
+    let ops = [a_ops.clone(), v_ops.clone(), t_read.clone()];
+    let combined_ops = DensePolynomial::merge(&ops);
+    let mems = [v_mem.clone(), t_final.clone()];
+    let combined_mem = DensePolynomial::merge(&mems);
+
+    Self {
+      a_ops,
+      v_ops,
+      v_mem,
+      t_read,
+      t_final,
+
+      address_offset,
+
+      combined_ops,
+      combined_mem,
+    }
+  }
+
+  pub fn commit<G: CurveGroup<ScalarField = F>>(
+    &self,
+  ) -> (ProgramCommitmentGens<G>, ProgramCommitment<G>) {
+    let gens = ProgramCommitmentGens::new(self.ops_size(), self.mem_size());
+
+    println!("committing ops");
+    let (ops_sized_commitments, _) = self.combined_ops.commit(&gens.gens_ops, None);
+    let ops_sized_commitments = CombinedTableCommitment::new(ops_sized_commitments);
+    println!("Committed ops");
+    let (mem_sized_commitments, _) = self.combined_mem.commit(&gens.gens_mem, None);
+    let mem_sized_commitments = CombinedTableCommitment::new(mem_sized_commitments);
+
+    let commitments = ProgramCommitment {
+      ops_sized_commitments,
+      mem_sized_commitments,
+    };
+
+    (gens, commitments)
+  }
+}
+
+pub struct PCProof<F: PrimeField> {
+  _marker: PhantomData<F>,
+}
+
+pub struct ProgramCommitment<G: CurveGroup> {
+  // Contains:
+  // - a_ops, v_ops (trace)
+  // - t_read
+  pub ops_sized_commitments: CombinedTableCommitment<G>,
+
+  /// Contains:
+  /// - v_mem (ELF)
+  /// - t_final
+  pub mem_sized_commitments: CombinedTableCommitment<G>,
+}
+
+pub struct ProgramCommitmentGens<G: CurveGroup> {
+  pub gens_ops: PolyCommitmentGens<G>,
+  pub gens_mem: PolyCommitmentGens<G>,
+}
+
+impl<G: CurveGroup> ProgramCommitmentGens<G> {
+  pub fn new(ops_size: usize, mem_size: usize) -> Self {
+    debug_assert!(utils::is_power_of_two(ops_size));
+    debug_assert!(utils::is_power_of_two(mem_size));
+
+    let num_vars_ops = (3 * ops_size).log_2();
+    let num_vars_mem = (2 * mem_size).log_2();
+
+    let gens_ops = PolyCommitmentGens::new(num_vars_ops, b"ops_commitment");
+    let gens_mem = PolyCommitmentGens::new(num_vars_mem, b"mem_commitment");
+
+    ProgramCommitmentGens { gens_ops, gens_mem }
+  }
+}
+
+impl<F: PrimeField> MemBatchInfo for PCPolys<F> {
+  fn ops_size(&self) -> usize {
+    self.a_ops.len()
+  }
+
+  fn mem_size(&self) -> usize {
+    self.v_mem.len()
+  }
+
+  fn num_memories(&self) -> usize {
+    1
+  }
+}
+
+impl<F: PrimeField> BGPCInterpretable<F> for PCPolys<F> {
+  fn a_ops(&self, memory_index: usize, leaf_index: usize) -> F {
+    debug_assert_eq!(memory_index, 0);
+    self.a_ops[leaf_index]
```

Comment:
We're performing multiple reads per element of `a_ops` (see `contiguous_reads_per_access` in `prove_program_code`) so this should be something like `self.a_ops[leaf_index / contiguous_reads_per_access] + leaf_index % contiguous_reads_per_access`

Reply:
actually now that I think about it, we should be using the same trick for the timestamps to avoid committing to duplicate elements

Reply:
Swapped for the multi-lane-v strategy.